### PR TITLE
Colorize child process output

### PR DESCRIPTION
**Pull request by holisticode**
Closes https://ava-labs.atlassian.net/browse/CHS-278

### CODE REVIEW

File: utils/color.go [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+	logging.Yellow,
+}
+
+type color struct {
+	lock       sync.Mutex
+	usedColors []logging.Color
+}
+
+// ColorPicker allows to assign a color to different clients
+var ColorPicker = &color{
```
Comment (danlaine) [inline utils/color.go:29]: Let's not use a global variable here. We could have a function `GetNextColor(i int)` which returns the color at `supportedColors[i%len(supportedColors)`. (That might be the easiest.) Or we could have a function `NewColorPicker` which returns a color picker struct.

File: local/network.go [hunk 2]
```diff
@@ -127,25 +128,58 @@ func init() {
...
+type nodeProcessCreator struct {
+	color logging.Color
+}
+
+// default variables pointing to the default functions to get the os output files
+// we can replace these in testing (see TestChildCmdRedirection in local/network_test.go)
+var (
+	stdOutFunc = getDefaultStdOut
+	stdErrFunc = getDefaultStdErr
+)
```
Comment (danlaine) [inline local/network.go:145]: I think we can avoid having the globals/overwriting them in tests by adding additional fields to `nodeProcessCreator`.

Namely, we can add fields `stdout io.Writer` and `stderr io.Writer` to control where stdout and stderr are written for testing

File: local/network.go [hunk 3]
```diff
@@ -127,25 +128,58 @@ func init() {
...
 }
 
+// NodeProcessCreator is an interface for new node process creation
 type NodeProcessCreator interface {
 	NewNodeProcess(config node.Config, args ...string) (NodeProcess, error)
 }
 
-type nodeProcessCreator struct{}
+type nodeProcessCreator struct {
+	color logging.Color
```
Comment (danlaine) [inline local/network.go:137]: comment please

File: utils/color.go [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+	go func(scanner *bufio.Scanner) {
+		// we should not need any go routine control here:
+		// when the program exits, `Scan()` will hit an EOF and return false,
+		// and therefore the routine terminates
+		for scanner.Scan() {
+			txt := assignedColor.Wrap(fmt.Sprintf("[%s] %s\n", wrapText, scanner.Text()))
+			if _, err := writer.Write([]byte(txt)); err != nil {
+				// TODO better handling required? Failing and returning doesn't seem quite right either
+				fmt.Println("failed to write wrapped text to writer")
+			}
```
Comment (danlaine) [inline utils/color.go:66]: I think we can just ignore errors on write since I don't think they should ever really happen

**Review by danlaine — CHANGES_REQUESTED** [review]

File: utils/color.go (hunk 1 above)
Comment (holisticode) [reply utils/color.go:29]: I think the use of a global variable is justified here. Specifically, the suggested change significantly deteriorates usage for clients. Also this is a fairly uncritical part of the code.

File: local/network.go (hunk 2 above)
Comment (holisticode) [reply local/network.go:145]: Much better and elegant, also uses way less code

File: utils/color.go (hunk 4 above)
Comment (holisticode) [reply utils/color.go:66]: I agree

**Pull request merged**